File: flutter/lib/src/webengage_payload_mapper.dart
import 'dart:convert';

import 'package:digia_engage/digia_engage.dart';
import 'package:we_personalization_flutter/we_personalization_flutter.dart';

/// Maps WebEngage callback payloads to Digia trigger payloads.
class WebEngagePayloadMapper {
  /// Maps a WE in-app notification data map (with injected `experimentId`)
  /// to a list of Digia [InAppPayload]s.
  ///
  /// Produces one inline payload when `type == "inline"`, or one nudge payload
  /// when `command` and `viewId` are present.
  List<InAppPayload> map(Map<String, dynamic> data) {
    final normalized = _normalizeWithDigiaContract(data);
    final campaignId = _str(
          normalized['experimentId'] ??
              normalized['id'] ??
              normalized['campaignId'],
        ) ??
        '';
    final screenId =
        _str(normalized['screenId'] ?? normalized['screen_id']) ?? '*';
    if (campaignId.isEmpty) {
      return const <InAppPayload>[];
    }

    final args = _normalizeArgs(normalized['args']);
    final payloads = <InAppPayload>[];

    final type = _str(normalized['type'])?.toLowerCase();
    final command = _str(normalized['command'])?.toUpperCase();
    final viewId = _str(normalized['viewId']) ?? '';
    final variationId = _str(normalized['variationId']) ?? '';

    if (type == 'inline') {
      final placementKey = _str(normalized['placementKey']) ?? '';
      if (placementKey.isNotEmpty && viewId.isNotEmpty) {
        payloads.add(
          InAppPayload(
            id: campaignId,
            content: <String, dynamic>{
              'type': 'inline',
              'screenId': screenId,
              'placementKey': placementKey,
              'viewId': viewId,
              'args': args,
            },
            cepContext: <String, dynamic>{
              'experimentId': campaignId,
              'campaignId': campaignId,
              if (variationId.isNotEmpty) 'variationId': variationId,
            },
          ),
        );
      }
    } else if (command != null && command.isNotEmpty && viewId.isNotEmpty) {
      payloads.add(
        InAppPayload(
          id: campaignId,
          content: <String, dynamic>{
            'command': command,
            'viewId': viewId,
            'screenId': screenId,
            'args': args,
          },
          cepContext: <String, dynamic>{'experimentId': campaignId},
        ),
      );
    }

    return payloads;
  }

  /// Maps a WE inline campaign data object to a single Digia [InAppPayload]
  /// for a [DigiaSlot] / placement to render.
  InAppPayload? mapInline(WECampaignData data) {
    final campaignId = data.campaignId ?? '';
    final targetViewId = data.targetViewId ?? '';
    if (campaignId.isEmpty || targetViewId.isEmpty) return null;

    final customData =
        Map<String, dynamic>.from(data.weCampaignContent?.customData ?? {});

    final screenId =
        _str(customData['screenId'] ?? customData['screen_id']) ?? '';
    final componentId = _str(customData['componentId']) ?? '';
    if (componentId.isEmpty) return null;
    final args = _normalizeArgs(customData['args']);

    final metadata = data.weCampaignContent?.customData ?? {};
    final variationId = _str(metadata['variationId']) ?? '';
    final propertyId = _str(metadata['propertyId']) ?? targetViewId;

    return InAppPayload(
      id: '$campaignId:$targetViewId',
      content: <String, dynamic>{
        'type': 'inline',
        'screenId': screenId,
        'placementKey': targetViewId,
        'viewId': componentId,
        'args': args,
      },
      cepContext: <String, dynamic>{
        'experimentId': campaignId,
        'campaignId': campaignId,
        'propertyId': propertyId,
        if (variationId.isNotEmpty) 'variationId': variationId,
      },
    );
  }

  String? _str(Object? value) {
    if (value == null) return null;
    final s = value.toString().trim();
    return s.isEmpty ? null : s;
  }

  Map<String, dynamic> _normalizeArgs(Object? value) {
    if (value is Map<String, dynamic>) return value;
    if (value is Map) {
      return <String, dynamic>{
        for (final e in value.entries)
          if (e.key is String) e.key as String: e.value,
      };
    }
    if (value is String) {
      final decoded = _tryDecodeJsonObject(value);
      return decoded ?? const <String, dynamic>{};
    }
    return const <String, dynamic>{};
  }

  Map<String, dynamic> _normalizeWithDigiaContract(Map<String, dynamic> raw) {
    final topLevel = _parseContractFromMap(raw, source: 'top_level');
    if (topLevel != null) return <String, dynamic>{...raw, ...topLevel};

    final htmlContract = _extractContractFromHtml(raw);
    if (htmlContract != null) {
      return <String, dynamic>{...raw, ...htmlContract};
    }

    return raw;
  }

  Map<String, dynamic>? _extractContractFromHtml(Map<String, dynamic> raw) {
    // Source 1: <digia> attribute-based tag (matches Android HtmlDigiaTagContractSource)
    final digiaTagRegex = RegExp(
      r'<digia\b([^/]*)/?>',
      caseSensitive: false,
    );
    for (final text in _collectTextCandidates(raw)) {
      for (final match in digiaTagRegex.allMatches(text)) {
        final attrs = _parseHtmlAttributes(match.group(1) ?? '');
        if (attrs.isEmpty) continue;
        final contract = _parseContractFromMap(attrs, source: 'html_digia_tag');
        if (contract != null) return contract;
      }
    }

    // Source 2: <script digia-payload> tag (legacy)
    final scriptRegex = RegExp(
      '<script[^>]*digia-payload[^>]*>(.*?)</script>',
      caseSensitive: false,
      dotAll: true,
    );
    for (final text in _collectTextCandidates(raw)) {
      for (final match in scriptRegex.allMatches(text)) {
        final body = _htmlUnescape(match.group(1)?.trim() ?? '');
        if (body.isEmpty) continue;
        final decoded = _tryDecodeJsonObject(body);
        if (decoded == null) continue;
        final contract = _parseContractFromMap(decoded, source: 'html_script');
        if (contract != null) return contract;
      }
    }

    return null;
  }

  /// Parses `key="value"` attribute pairs from HTML tag attribute string.
  /// Normalizes kebab-case / underscore_case attribute names to camelCase,
  /// matching Android's `DigiaTagParser.mapAttributes` behaviour.
  Map<String, dynamic> _parseHtmlAttributes(String attrsString) {
    final result = <String, dynamic>{};
    final attrRegex =
        RegExp(r'''(\w[\w-]*)(?:\s*=\s*(?:"([^"]*)"|'([^']*)'|(\S+)))?''');
    for (final match in attrRegex.allMatches(attrsString)) {
      final rawKey = match.group(1) ?? '';
      final value = match.group(2) ?? match.group(3) ?? match.group(4) ?? '';
      if (rawKey.isEmpty) continue;
      final key = _normalizeAttrKey(rawKey);
      result[key] = _htmlUnescape(value);
    }
    return result;
  }

  String _normalizeAttrKey(String key) {
    switch (key) {
      case 'view-id':
      case 'view_id':
        return 'viewId';
      case 'placement-key':
      case 'placement_key':
      case 'placement':
        return 'placementKey';
      case 'screen-id':
      case 'screen_id':
        return 'screenId';
      default:
        return key;
    }
  }

  Map<String, dynamic>? _parseContractFromMap(
    Map<String, dynamic> raw, {
    required String source,
  }) {
    final type = _str(raw['type'])?.toLowerCase();
    final command = _normalizeCommand(_str(raw['command']));
    final resolvedCommand = type == 'inline' ? null : command;

    // At least one routing signal must exist
    if (type == null && resolvedCommand == null) return null;

    final viewId = _str(raw['viewId']);
    if (viewId == null) return null;

    final placementKey = _str(raw['placementKey']);

    // Inline requires placementKey
    if (type == 'inline' && placementKey == null) return null;

    final screenId = _str(raw['screenId'] ?? raw['screen_id']);
    return <String, dynamic>{
      if (type != null) 'type': type,
      if (resolvedCommand != null) 'command': resolvedCommand,
      'viewId': viewId,
      'args': _normalizeArgs(raw['args']),
      'digiaContractSource': source,
      if (placementKey != null) 'placementKey': placementKey,
      if (screenId != null) 'screenId': screenId,
    };
  }

  String? _normalizeCommand(String? value) {
    final normalized = value?.toUpperCase();
    if (normalized == 'SHOW_DIALOG' || normalized == 'SHOW_BOTTOM_SHEET') {
      return normalized;
    }
    return null;
  }

  List<String> _collectTextCandidates(Object? node) {
    final out = <String>[];
    final queue = <Object?>[node];
    while (queue.isNotEmpty) {
      final current = queue.removeAt(0);
      if (current is String) {
        if (current.contains('<')) out.add(current);
      } else if (current is Map) {
        queue.addAll(current.values);
      } else if (current is List) {
        queue.addAll(current);
      }
    }
    return out;
  }

  Map<String, dynamic>? _tryDecodeJsonObject(String value) {
    try {
      final decoded = jsonDecode(value);
      if (decoded is Map<String, dynamic>) return decoded;
      if (decoded is Map) {
        return <String, dynamic>{
          for (final e in decoded.entries)
            if (e.key is String) e.key as String: e.value,
        };
      }
      return null;
    } on FormatException {
      return null;
    }
  }

  String _htmlUnescape(String value) {
    return value
        .replaceAll('&quot;', '"')
        .replaceAll('&#34;', '"')
        .replaceAll('&apos;', "'")
        .replaceAll('&#39;', "'")
        .replaceAll('&lt;', '<')
        .replaceAll('&gt;', '>')
        .replaceAll('&amp;', '&');
  }
}

File: react-native/src/WebEngagePayloadMapper.ts
import type { InAppPayload } from './types';
import { SuppressionMode, defaultConfig } from './config';
import type { WebEngagePluginConfig } from './config';

// ─── Public API ───────────────────────────────────────────────────────────────

/**
 * Maps WebEngage callback payloads to Digia [InAppPayload]s.
 *
 * Mirrors the logic in the Android `WebEngagePayloadMapper` and the Flutter
 * `WebEngagePayloadMapper` exactly:
 *
 * 1. Extract a Digia contract from the raw data (top-level keys → `<digia>`
 *    HTML attribute tag → `<script digia-payload>` tag).
 * 2. If `type == "inline"`, build an inline payload (requires `placementKey`
 *    and `viewId`).
 * 3. If `command` is present, build a nudge payload.
 * 4. Otherwise fall back to a forced-dialog payload when
 *    `suppressionMode == SUPPRESS_ALL` and `forcedDialogComponentId` is set.
 */
export class WebEngagePayloadMapper {
    private readonly _config: WebEngagePluginConfig;

    constructor(config: Partial<WebEngagePluginConfig> = {}) {
        this._config = { ...defaultConfig, ...config };
    }

    /**
     * Maps a WebEngage in-app notification data map (with injected
     * `experimentId`) to a list of Digia [InAppPayload]s.
     */
    map(data: Record<string, unknown>): InAppPayload[] {
        const normalized = this._normalizeWithDigiaContract(data);

        const campaignId = str(
            normalized['experimentId'] ??
            normalized['id'] ??
            normalized['campaignId'],
        );
        if (!campaignId) return [];

        const screenId =
            str(normalized['screenId'] ?? normalized['screen_id']) ?? '*';
        const args = normalizeArgs(normalized['args']);
        const payloads: InAppPayload[] = [];

        const type = str(normalized['type'])?.toLowerCase();
        const command = normalizeCommand(str(normalized['command']));
        const viewId = str(normalized['viewId']) ?? '';
        const variationId = str(normalized['variationId']) ?? '';

        if (type === 'inline') {
            const placementKey = str(normalized['placementKey']) ?? '';
            if (placementKey && viewId) {
                payloads.push({
                    id: campaignId,
                    content: {
                        type: 'inline',
                        screenId,
                        placementKey,
                        componentId: viewId,
                        args,
                    },
                    cepContext: {
                        experimentId: campaignId,
                        campaignId,
                        ...(variationId && { variationId }),
                    },
                });
            }
        } else if (command && viewId) {
            payloads.push({
                id: campaignId,
                content: {
                    command,
                    viewId,
                    screenId,
                    args,
                },
                cepContext: { experimentId: campaignId },
            });
        } else {
            const forcedId = this._config.forcedDialogComponentId?.trim() ?? '';
            if (
                this._config.suppressionMode === SuppressionMode.SUPPRESS_ALL &&
                forcedId
            ) {
                payloads.push(buildForcedDialogPayload(campaignId, forcedId));
            }
        }

        return payloads;
    }

    // ─── Contract extraction ────────────────────────────────────────────────────

    private _normalizeWithDigiaContract(
        raw: Record<string, unknown>,
    ): Record<string, unknown> {
        const topLevel = parseContractFromMap(raw, 'top_level');
        if (topLevel) return { ...raw, ...topLevel };

        const htmlContract = extractContractFromHtml(raw);
        if (htmlContract) return { ...raw, ...htmlContract };

        return raw;
    }
}

// ─── Contract extraction helpers (module-private) ────────────────────────────

function extractContractFromHtml(
    raw: Record<string, unknown>,
): Record<string, unknown> | null {
    const candidates = collectTextCandidates(raw);

    // Source 1: <digia> attribute-based tag (matches Android HtmlDigiaTagContractSource)
    const digiaTagRe = /<digia\b([^/]*)\/?>/gi;
    for (const text of candidates) {
        for (const match of text.matchAll(digiaTagRe)) {
            const attrs = parseHtmlAttributes(match[1] ?? '');
            if (Object.keys(attrs).length === 0) continue;
            const contract = parseContractFromMap(attrs, 'html_digia_tag');
            if (contract) return contract;
        }
    }

    // Source 2: <script digia-payload> tag (legacy)
    const scriptRe = /<script[^>]*digia-payload[^>]*>([\s\S]*?)<\/script>/gi;
    for (const text of candidates) {
        for (const match of text.matchAll(scriptRe)) {
            const body = htmlUnescape((match[1] ?? '').trim());
            if (!body) continue;
            const decoded = tryDecodeJsonObject(body);
            if (!decoded) continue;
            const contract = parseContractFromMap(decoded, 'html_script');
            if (contract) return contract;
        }
    }

    return null;
}

function parseContractFromMap(
    raw: Record<string, unknown>,
    source: string,
): Record<string, unknown> | null {
    const type = str(raw['type'])?.toLowerCase();
    const resolvedCommand = type === 'inline' ? null : normalizeCommand(str(raw['command']));

    // At least one routing signal must exist
    if (!type && !resolvedCommand) return null;

    const viewId = str(raw['viewId']);
    if (!viewId) return null;

    const placementKey = str(raw['placementKey']);

    // Inline requires placementKey
    if (type === 'inline' && !placementKey) return null;

    const screenId = str(raw['screenId'] ?? (raw['screen_id']));

    const contract: Record<string, unknown> = {
        ...(type && { type }),
        ...(resolvedCommand && { command: resolvedCommand }),
        viewId,
        args: normalizeArgs(raw['args']),
        digiaContractSource: source,
        ...(placementKey && { placementKey }),
        ...(screenId && { screenId }),
    };
    return contract;
}

function parseHtmlAttributes(
    attrsString: string,
): Record<string, unknown> {
    const result: Record<string, unknown> = {};
    // Matches: key="value" | key='value' | key=value | standalone key
    const attrRe = /(\w[\w-]*)(?:\s*=\s*(?:"([^"]*)"|'([^']*)'|(\S+)))?/g;
    for (const match of attrsString.matchAll(attrRe)) {
        const key = match[1];
        const value = match[2] ?? match[3] ?? match[4] ?? '';
        if (key) result[key] = htmlUnescape(value);
    }
    return result;
}

// ─── Payload builders ─────────────────────────────────────────────────────────

function buildForcedDialogPayload(
    campaignId: string,
    componentId: string,
): InAppPayload {
    return {
        id: `${campaignId}:forced_dialog`,
        content: {
            command: 'SHOW_DIALOG',
            viewId: componentId,
            screenId: '*',
            args: {},
        },
        cepContext: { experimentId: campaignId },
    };
}

// ─── Utilities ────────────────────────────────────────────────────────────────

function str(value: unknown): string | null {
    if (value == null) return null;
    const s = String(value).trim();
    return s.length > 0 ? s : null;
}

function normalizeCommand(value: string | null): string | null {
    const upper = value?.toUpperCase();
    if (upper === 'SHOW_DIALOG' || upper === 'SHOW_BOTTOM_SHEET') return upper;
    return null;
}

function normalizeArgs(value: unknown): Record<string, unknown> {
    if (value == null) return {};
    if (typeof value === 'object' && !Array.isArray(value)) {
        return value as Record<string, unknown>;
    }
    if (typeof value === 'string') {
        return tryDecodeJsonObject(value) ?? {};
    }
    return {};
}

function collectTextCandidates(node: unknown): string[] {
    const results: string[] = [];
    const queue: unknown[] = [node];
    while (queue.length > 0) {
        const current = queue.shift();
        if (typeof current === 'string') {
            if (current.includes('<')) results.push(current);
        } else if (current != null && typeof current === 'object') {
            if (Array.isArray(current)) {
                queue.push(...current);
            } else {
                queue.push(...Object.values(current as Record<string, unknown>));
            }
        }
    }
    return results;
}

function tryDecodeJsonObject(
    value: string,
): Record<string, unknown> | null {
    try {
        const decoded: unknown = JSON.parse(value);
        if (
            decoded != null &&
            typeof decoded === 'object' &&
            !Array.isArray(decoded)
        ) {
            return decoded as Record<string, unknown>;
        }
        return null;
    } catch {
        return null;
    }
}

function htmlUnescape(value: string): string {
    return value
        .replace(/&quot;/g, '"')
        .replace(/&#34;/g, '"')
        .replace(/&apos;/g, "'")
        .replace(/&#39;/g, "'")
        .replace(/&lt;/g, '<')
        .replace(/&gt;/g, '>')
        .replace(/&amp;/g, '&');
}
